eth/downloader/queue.go

```diff
@@ -210,96 +310,160 @@ func (q *queue) Insert(headers []*types.Header, from uint64) []*types.Header {
 			glog.V(logger.Warn).Infof("Header #%v [%x] broke chain ancestry", header.Number, hash[:4])
 			break
 		}
-		// Queue the header for body retrieval
+		// Make sure no duplicate requests are executed
+		if _, ok := q.blockTaskPool[hash]; ok {
+			glog.V(logger.Warn).Infof("Header #%d [%x] already scheduled for block fetch", header.Number.Uint64(), hash[:4])
+			continue
+		}
+		if _, ok := q.receiptTaskPool[hash]; ok {
+			glog.V(logger.Warn).Infof("Header #%d [%x] already scheduled for receipt fetch", header.Number.Uint64(), hash[:4])
+			continue
+		}
+		// Queue the header for content retrieval
+		q.blockTaskPool[hash] = header
+		q.blockTaskQueue.Push(header, -float32(header.Number.Uint64()))
```

Comment:
Ah, OK. I probably disagree with that design decision - floats are evil and should be avoided unless there are very good reasons for them :) But in this case it shouldn't be a problem given the number of queue items we'll see in practice.